test/type-error.js

```diff
@@ -0,0 +1,105 @@
+import test from 'ava';
+import avaRuleTester from 'eslint-ava-rule-tester';
+import rule from '../rules/type-error';
+
+const ruleTester = avaRuleTester(test, {
+	env: {
+		es6: true
+	}
+});
+
+const errors = [{
+	ruleId: 'type-error',
+	message: '`new Error()` is too unspecific for a typecheck, use `new TypeError()` instead.'
+}];
+
+ruleTester.run('type-error', rule, {
+	valid: [
+		`if (Array.isArray(foo) || ArrayBuffer.isView(foo)) {
+			throw new TypeError();
+		}`,
+		`if (!isFinite(foo)) {
+			throw new TypeError();
+		}`,
+		`if (isNaN(foo)) {
+			throw new TypeError();
+		}`,
+		`if (foo instanceof boo) {
+			throw new TypeError();
+		}`,
+		`if (typeof boo === 'Boo') {
+			throw new TypeError();
+		}`,
+		`if (Number.isNaN(foo)) {
+			throw new TypeError();
+		}`,
+		`if (Number.isFinite(foo) && Number.isSafeInteger(foo) && Number.isInteger(foo)) {
+			throw new TypeError();
+		}`,
+		`if (Array.isArray(foo) || (Blob.isBlob(foo) || Blip.isBlip(foo))) {
+			throw new TypeError();
+		}`,
+		`if (typeof foo === 'object' || (Object.isFrozen(foo) || 'String' === typeof foo)) {
+			throw new TypeError();
+		}`,
+		`if (Object.isSealed(foo) && _.isArguments(foo)) {
+			throw new TypeError();
+		}`
```

Comment:
According to the docs, this should be valid as well

```js
if (!Number.isNaN(foo) && foo === 10) {
    throw new Error();
}
```

Because all the expressions should be a typecheck before this rule fails. Maybe it's a good idea to add it something like this to the valid tests.